frame/scheduler/src/lib.rs

```diff
@@ -240,21 +226,37 @@ pub mod pallet {
 		type WeightInfo: WeightInfo;
 
 		/// The preimage provider with which we look up call hashes to get the call.
-		type PreimageProvider: PreimageProviderAndMaybeRecipient<Self::Hash>;
-
-		/// If `Some` then the number of blocks to postpone execution for when the item is delayed.
-		type NoPreimagePostponement: Get<Option<Self::BlockNumber>>;
+		type Preimages: QueryPreimage + StorePreimage;
 	}
 
-	/// Items to be executed, indexed by the block number that they should be executed on.
 	#[pallet::storage]
-	pub type Agenda<T: Config> =
-		StorageMap<_, Twox64Concat, T::BlockNumber, Vec<Option<ScheduledV3Of<T>>>, ValueQuery>;
+	pub type IncompleteSince<T: Config> = StorageValue<_, T::BlockNumber>;
 
-	/// Lookup from identity to the block number and index of the task.
+	/// Items to be executed, indexed by the block number that they should be executed on.
+	#[pallet::storage]
+	pub type Agenda<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		T::BlockNumber,
+		BoundedVec<Option<ScheduledOf<T>>, T::MaxScheduledPerBlock>,
```

Comment:
Why did you opt to keep the `Option` logic here?  
We then have to ensure that the cancel and rescheduling logic always removes `None`-only agendas from storage.

Reply:
We allow tasks to be referenced (and later removed) by the `Address` type, which is the block number & Agenda index. We could refactor away from it, but I don't see much point.